src/utils/revision-utils.js

```diff
@@ -0,0 +1,125 @@
+import deepEqual from 'fast-deep-equal';
+import permissions, { hasUserPermission } from '../domain/permissions.js';
+import { isRoomOwner, isRoomOwnerOrInvitedCollaborator } from './room-utils.js';
+
+export function checkRevisionOnDocumentCreation({ newRevision, room, user }) {
+  if (!hasUserPermission(user, permissions.EDIT_DOC)) {
+    throw new Error('User is not allowed to create documents');
+  }
+
+  const { publicContext, roomContext } = newRevision;
+
+  if (room) {
+    const userIsRoomOwner = isRoomOwner({ room, userId: user._id });
+    const userIsRoomOwnerOrInvitedCollaborator = isRoomOwnerOrInvitedCollaborator({ room, userId: user?._id });
+
+    if (roomContext.draft && !userIsRoomOwner) {
+      throw new Error('Only room owners can create a draft document');
+    }
+
+    if (!userIsRoomOwnerOrInvitedCollaborator) {
+      throw new Error('Only room owners or collaborators can create a room document');
+    }
+  }
+
+  if (!room) {
+    const userCanAssignAnyAccreditedEditor = hasUserPermission(user, permissions.MANAGE_ACCREDITED_EDITORS);
+    const userCanAssignSelfAsAccreditedEditor = hasUserPermission(user, permissions.PROTECT_OWN_DOC);
+    const userCanProtectOwnNewDoc = hasUserPermission(user, permissions.PROTECT_OWN_DOC);
+    const userCanProtectAnyDoc = hasUserPermission(user, permissions.PROTECT_ANY_DOC);
+    const userCanArchiveDoc = hasUserPermission(user, permissions.ARCHIVE_DOC);
+    const userCanVerifyDoc = hasUserPermission(user, permissions.VERIFY_DOC);
+    const userCanReviewDoc = hasUserPermission(user, permissions.REVIEW_DOC);
+
+    const { accreditedEditors } = publicContext;
+    if (accreditedEditors.length) {
+      const userIsOnlyAccreditedEditor = accreditedEditors.length === 1 && accreditedEditors[0] === user._id;
+
+      if (!userIsOnlyAccreditedEditor && !userCanAssignAnyAccreditedEditor) {
+        throw new Error('User is not allowed to assign accredited editors');
+      }
+
+      if (userIsOnlyAccreditedEditor && !userCanAssignAnyAccreditedEditor && !userCanAssignSelfAsAccreditedEditor) {
+        throw new Error('User is not allowed to assign themselves as accredited editor');
+      }
```

Comment:
Took me a while to get these cases somehow, but it's clear now. Which only shows how important it is to have this logic documented here this way through this validation step..